data/summary/types/summary.go

```diff
@@ -0,0 +1,284 @@
+package types
+
+import (
+	glucoseDatum "github.com/tidepool-org/platform/data/types/blood/glucose"
+	insulinDatum "github.com/tidepool-org/platform/data/types/insulin"
+	"github.com/tidepool-org/platform/errors"
+	"github.com/tidepool-org/platform/pointer"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"time"
+)
+
+const (
+	SummaryTypeCGM = "cgm"
+	SummaryTypeBGM = "bgm"
+
+	lowBloodGlucose      = 3.9
+	veryLowBloodGlucose  = 3.0
+	highBloodGlucose     = 10.0
+	veryHighBloodGlucose = 13.9
+	summaryGlucoseUnits  = "mmol/L"
+	hoursAgoToKeep       = 30 * 24
+)
+
+type BucketData interface {
+	CGMBucketData | BGMBucketData
+}
+
+type RecordTypes interface {
+	glucoseDatum.Glucose | insulinDatum.Insulin
+}
+
+type RecordTypesPt[T RecordTypes] interface {
+	*T
+	GetTime() *time.Time
+}
+
+// Glucose reimplementation with only the fields we need, to avoid inheriting Base, which does
+// not belong in this collection
+type Glucose struct {
+	Units string  `json:"units" bson:"units"`
+	Value float64 `json:"value" bson:"value"`
+}
+
+type UserLastUpdated struct {
+	LastData   time.Time
+	LastUpload time.Time
+}
+
+type Config struct {
+	SchemaVersion int `json:"schemaVersion" bson:"schemaVersion"`
+
+	// these are just constants right now.
+	HighGlucoseThreshold     float64 `json:"highGlucoseThreshold" bson:"highGlucoseThreshold"`
+	VeryHighGlucoseThreshold float64 `json:"veryHighGlucoseThreshold" bson:"veryHighGlucoseThreshold"`
+	LowGlucoseThreshold      float64 `json:"lowGlucoseThreshold" bson:"lowGlucoseThreshold"`
+	VeryLowGlucoseThreshold  float64 `json:"VeryLowGlucoseThreshold" bson:"VeryLowGlucoseThreshold"`
+}
+
+type Dates struct {
+	// date tracking
+	HasLastUploadDate bool       `json:"hasLastUploadDate" bson:"hasLastUploadDate"`
+	LastUploadDate    time.Time  `json:"lastUploadDate" bson:"lastUploadDate"`
+	LastUpdatedDate   time.Time  `json:"lastUpdatedDate" bson:"lastUpdatedDate"`
+	FirstData         time.Time  `json:"firstData" bson:"firstData"`
+	LastData          *time.Time `json:"lastData" bson:"lastData"`
+	OutdatedSince     *time.Time `json:"outdatedSince" bson:"outdatedSince"`
+}
+
+type Bucket[T BucketData, S BucketDataPt[T]] struct {
+	Date           time.Time `json:"date" bson:"date"`
+	LastRecordTime time.Time `json:"lastRecordTime" bson:"lastRecordTime"`
+
+	Data S
+}
+
+type BucketDataPt[T BucketData] interface {
+	*T
+	CalculateStats(interface{}, *time.Time) error
+}
+
+func CreateBucket[T BucketData, A BucketDataPt[T]](t time.Time) *Bucket[T, A] {
+	bucket := new(Bucket[T, A])
+	bucket.Date = t
+	return bucket
+}
+
+type Buckets[T BucketData, S BucketDataPt[T]] []Bucket[T, S]
```

Comment:
This may fall more into personal preference, but can we "flatten" the constraints a bit? The fewer type constraints and parameters in a signature the simpler it is to read and understand IMO.

I see things like `BucketData` and `BucketDataPt` always(?) being used together, could we merge them? Is there any place else in the code that might prevent this that I'm missing?

For example:
One type constraint `BucketDataPt` instead of both `BucketData` and `BucketDataPt`:
```
type BucketDataPt interface {
	*CGMBucketData | *BGMBucketData
	CalculateStats(interface{}, *time.Time) error
}
```

This would simplify signatures and invocations of some of the generic functions and types.

```go
func CreateBucket[T BucketData, A BucketDataPt[T]]
// becomes
func CreateBucket[T BucketDataPt]

func AddBin[T BucketData, A BucketDataPt[T], S Buckets[T, A]]
// becomes
func AddBin[T BucketDataPt, S Buckets[T]]

func AddData[T BucketData, A BucketDataPt[T], S Buckets[T, A], ...]
// becomes
func AddData[T BucketDataPt, S Buckets[T], ...]

//It would also avoid somewhat confusing instantiations such as
Buckets Buckets[BGMBucketData, *BGMBucketData]
// becomes
Buckets[*BGMBucketData] 
```

The same can apply to `StatsPt` and `Stats` and anywhere else I've missed.

Reply:
This won't work in go 1.19 (I don't think it has changed in 1.20 either). Both type constraints are required in order to be able to **instantiate** the underlying struct.

Consider the following simplified example which demonstrates why your suggestion doesn't work:
```go
package main

import "fmt"

type BGMBucketData struct {
	Value int
}

func (b *BGMBucketData) Print() {
	fmt.Println("print was called")
}

func (b *BGMBucketData) SetValue(v int) {
	b.Value = v
}

type BucketData interface {
	*BGMBucketData
	SetValue(int)
	Print()
}

type Data interface {
	BGMBucketData
}

type DataPt[T Data] interface {
	*T
	SetValue(int)
	Print()
}

func SimpleCreate[T BucketData]() T {
	return *new(T) // *T is a pointer to the type constraint, dereference it 
}

func ComplexCreate[S DataPt[T], T Data]() S {
	return S(new(T)) // explicit casting is not required, but illustrates why the return signature works
}

func main() {
	var bucketA *BGMBucketData
	var bucketB *BGMBucketData
	bucketA = SimpleCreate[*BGMBucketData]()
	bucketB = ComplexCreate[*BGMBucketData, BGMBucketData]() // ComplexCreate[*BGMBucketData]() also works, because the second type can be inferred by the compiler

	bucketA.Print()            // works, prints "print was called"
	bucketB.Print()            // works, prints "print was called"
	bucketB.SetValue(5)        // works, sets Value of the struct
	fmt.Println(bucketB.Value) // prints 5
	bucketA.SetValue(5)        // nil pointer deref, because bucket A is nil
}
```

This is a limitation of Go generics. You can read more about it in the [design proposal](https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md#pointer-method-example).

Reply:
Forgot to add a playground link https://go.dev/play/p/bEBspzX6VGu?v=goprev

Reply:
Good to know, wasn't aware of that.

Reply:
@toddkazakov @Roukoswarf There is actually a way to do this using reflection, but that could also be just as confusing. More of an FYI.